crates/orderbook/src/order_validation.rs

```diff
@@ -497,6 +507,16 @@ async fn get_quote_and_check_fee(
     Ok(quote)
 }
 
+/// Checks whether or not an order is a market order.
+///
+/// A market order is defined as an order whose price is equal or worse to the
+/// best on-chain price that can be found at the time. This allows the protocol
+/// to detect orders that place orders that aren't intended to be traded right
+/// away and flag them as liquidity orders.
+fn is_market_priced_order(order: &OrderData, quote: &Quote) -> bool {
+    order.sell_amount.full_mul(quote.buy_amount) >= quote.sell_amount.full_mul(order.buy_amount)
+}
```

Comment:
The implementation of this function looks wrong because it doesn't check the order kind (sell vs buy). Indeed in the new test your buy orders are still OrderKind::Sell probably because of a copy paste error. I think the reason this works out is because one of (sell|buy)_amount is always guaranteed to match the same field in quote but I'm not sure. I feel it would be better to implement this with a match and a debug assert that this expectation is correct.

Reply:
I saw you just fixed the test but it is still surprising that the test did pass when this was wrong, right? This is why I would like a debug_assert so the wrong test would panic.

Reply:
> test did pass when this was wrong, right?

The `kind` was there for informational purposes and not because it was needed. If you think this is counter-intuitive and we should use stricter checks, that is fine with me. In fact, the two sets of tests for sell and buy orders were just to prove to myself that I got the price math correct (prices are hard 🙈), but the logic is really independent of whether or not amounts are respected (i.e. is the order limit price above or below the market price).

I don't have strong opinions though. If you feel the code is unclear, I would rather change it to have stricter checks or include additional documentation to clear that up that leave it as is though.

Reply:
On second thought you are right. Only the price matters. I thought about it wrong.

Reply:
I changed the doc comment to make this more explicit then. So we don't forget and have the same confusion a few months down the line.